libzerocash/Coin.cpp

```diff
@@ -22,6 +22,48 @@ Coin::Coin(): addr_pk(), cm(), rho(rho_size), r(zc_r_size), coinValue(v_size) {
 
 }
 
+Coin::Coin(const std::string bucket, Address& addr): addr_pk(), cm(), rho(rho_size), r(zc_r_size), coinValue(v_size) {
+    // Retreive and decode the private key
+    ECIES<ECP>::PrivateKey decodedPrivateKey;
+    decodedPrivateKey.Load(StringStore(addr.getEncryptionSecretKey()).Ref());
+
+    // Create the decryption session
+    AutoSeededRandomPool prng;
+    ECIES<ECP>::Decryptor decrypt(decodedPrivateKey);
+
+    // Convert the input string into a vector of bytes
+    std::vector<byte> bucket_bytes(bucket.begin(), bucket.end());
+
+    // Construct a temporary object to store the plaintext, large enough
+    // to store the plaintext if it were extended beyond the real size.
+    std::vector<unsigned char> plaintext;
+    // Size as needed, filling with zeros.
+    plaintext.resize(decrypt.MaxPlaintextLength(decrypt.CiphertextLength(v_size + zc_r_size + rho_size)), 0);
+
+    // Perform the decryption
+    decrypt.Decrypt(prng,
+                    &bucket_bytes[0],
+                    decrypt.CiphertextLength(v_size + zc_r_size + rho_size),
```

Comment:
`bucket_bytes.size()` is not necessarily the ciphertext length for our desired plaintext. If the user supplies an arbitrarily sized bucket it could overflow the plaintext vector storage, which we're only reserving enough for our plaintext value.